### ISSUE TITLE
MemNN padding issue

### ISSUE THREAD

**Issue by lematt1991** [facebookresearch/ParlAI#1649]
I'm seeing different results with MemNN while varying batch size.  The problem seems to relate to attention and padding.

If I have a memory that gets padded (due to another training sample in its batch being longer), we add 0's to the end of the memory so that their dimensions match.  When we [do attention](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/memnn/modules.py#L203-L204) we don't treat these paddings any differently which will lead to different results in `probs` after applying the softmax.  For example:

```Python
>>> F.softmax(torch.Tensor([1, 4, 5, 3]))
tensor([0.0120, 0.2418, 0.6572, 0.0889])
>>> F.softmax(torch.Tensor([1, 4, 5, 3, 0, 0, 0]))
tensor([0.0119, 0.2386, 0.6486, 0.0878, 0.0044, 0.0044, 0.0044])
>>>
```

This causes the results to diverge at each hop.  [Fairseq gets around this](https://github.com/pytorch/fairseq/blob/master/fairseq/modules/multihead_attention.py#L186-L189) by setting padded values to `-inf`

```Python
>>> F.softmax(torch.Tensor([1, 4, 5, 3, float('-inf'), float('-inf'), float('-inf')]))
tensor([0.0120, 0.2418, 0.6572, 0.0889, 0.0000, 0.0000, 0.0000])
```

**Comment by lematt1991** [comment]
Another thing to keep in mind is that [we add time features to padded memories](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/memnn/memnn.py#L191).  We'll probably want to fix this as well.  I'm happy to submit a PR, but maybe we should discuss further.

**Comment by stephenroller** [comment]
Historically MemNN is only trained using hogwild using batch size of 1. I'm not surprised there are padding issues. A PR would be greatly appreciated (and a test!)
